readthedocs/subscriptions/tests/test_event_handlers.py

```diff
@@ -0,0 +1,283 @@
+from unittest import mock
+
+from django.test import TestCase, override_settings
+from django.utils import timezone
+from django_dynamic_fixture import get
+from djstripe import models as djstripe
+from djstripe.enums import SubscriptionStatus
+
+from readthedocs.organizations.models import Organization
+from readthedocs.subscriptions import event_handlers
+from readthedocs.subscriptions.models import Plan, Subscription
+
+
+# Avoid signature verification.
+@override_settings(
+    RTD_ALLOW_ORGANIZATIONS=True,
+    STRIPE_WEBHOOK_SECRET="1234",
+    RTD_ORG_DEFAULT_STRIPE_SUBSCRIPTION_PRICE="trialing",
+)
+class TestEventHandlers(TestCase):
+
+    """Tests for Stripe API endpoint."""
+
+    def setUp(self):
+        self.organization = get(Organization, slug="org", email="test@example.com")
+        get(Plan, stripe_id="trialing", slug="trialing")
+
+    def test_subscription_updated_event(self):
+        """Test handled event."""
+        start_date = timezone.now()
+        end_date = timezone.now() + timezone.timedelta(days=30)
+        stripe_subscription = get(
+            djstripe.Subscription,
+            id="sub_9LtsU02uvjO6Ed",
+            status=SubscriptionStatus.active,
+            current_period_start=start_date,
+            current_period_end=end_date,
+            trial_end=end_date,
+        )
+        event = get(
+            djstripe.Event,
+            data={
+                "object": {
+                    "id": "sub_9LtsU02uvjO6Ed",
+                    "object": "subscription",
+                }
+            },
+        )
+        subscription = get(
+            Subscription,
+            stripe_id=stripe_subscription.id,
+            status=SubscriptionStatus.trialing,
+        )
+        event_handlers.update_subscription(event=event)
```

Comment:
Can we have at least one test that hits the `/api/v2/stripe/` URL (or whatever it is the new one)? I think it's fine to test our handler functions directly, but I'd like to keep testing the URL as well to know that it's properly set up and connected to the handlers, and we don't remove it by mistake.

I'm not sure if we remove them all o not. If we don't, can you point me to the code where they are?

Reply:
We used to have them at https://github.com/readthedocs/readthedocs-corporate/pull/1497/files#diff-4b62929ef6f21bfdcdbc2ae4f92bb356dae58c303a5e5cdb8299573f66f5700bL170

Reply:
So, I wanted to keep the tests at the top level, but djstripe will execute their handlers too, and those will try to re-sync all models hitting the stripe api, that will require mock a lot of things, to keep it simple I just decided to test the handlers directly. And just tried to write a test and see what could be mocked, but there are a lot of things going on before it hits the handlers, ended up testing the even registration instead https://github.com/readthedocs/readthedocs.org/pull/9651/commits/541878f103b88ce924828b5598a22ca3005d641c

Reply:
I'm not sure I understand what the event registration test does. Does this hit a URL managed by our application? That was my intention here. Test that we can hit that URL and it's registered in our application.

If it's too complex to do this, I'm fine skipping it. I thought it was worth a try, tho.